### DISCUSSION TITLE
Issue of nextjs + material-ui generate different class names

### DISCUSSION THREAD

**Discussion by codyng** [vercel/next.js#10898] [Help]
Hi,
I don't know what is the problem.
When I use material-ui in _app.js with:
```
const generateClassName = createGenerateClassName({
    disableGlobal: true,
    productionPrefix: 'someprefix'
});
<StylesProvider generateClassName={generateClassName}>......</StylesProvider>
```

and with Next.js's MyDocument.getInitialProps():

```
    const sheets = new ServerStyleSheets({
        serverGenerateClassName: generateClassName
    });
```

It is because I have lots of pages defined in exportPathMap.
I found that the classes name, e.g. the class name used in the header bar <header> is firstly kept the same even I browse another page. However, when I refresh the page (Cmd + R), that class name is different and so the page style is crashed. The incremented number is very very large. From e.g. someprefix5 -> someprefix245.

What is the problem inside? Is it the configuration issue? I don't know if I need to configure it in material-ui side or next.js side.

Any help is highly appreciated! Thanks!

**Comment by waptik** [comment]
Hey @codyng , kindly check this out https://github.com/mui-org/material-ui/tree/master/examples/nextjs
You can use it as a boilerplate then build upon it.

**Comment by codyng** [comment]
Thanks but I have referred it. The code is not used with StylesProvider, productionPrefix and exportPathMap in nextjs config. I would like to point out the class names are inconsistent if enabling the productionPrefix or disableGlobal. If I disable all of them, i.e. pass no parameter to <StyleProvider> and take out the serverGenerateClassName in ServerStyleSheets() constructor, everything is fine.

Any idea?

**Comment by nazarien** [comment]
Hello @codyng,
have you found the fix?

**Comment by nazarien** [comment]
I've found why it was not working. `createGenerateClassName` instance should be initialized in `getInitialProps`. So if you declare `createGenerateClassName` object  outside `getInitialProps` func (even it's not the same instance what we have on client side) we will get that issue which will increment `productionPrefix` each page reload.

It won't work:
```
const generateClassName = createGenerateClassName({
  productionPrefix: 'pdp',
});

export default class Document extends NextDocument {
  static async getInitialProps(ctx) {
    const styledComponentSheet = new StyledComponentSheets();
    const materialUiSheets = new MaterialUiServerStyleSheets({
      serverGenerateClassName: generateClassName,
    });
...
```

You should move `generateClassName` inside `getInitialProps` (or use the way mentioned here with function https://github.com/mui-org/material-ui/issues/17423#issuecomment-680072139).

It's not clear from the docs
